rasa_core/policies/torch_policy.py

```diff
@@ -0,0 +1,234 @@
+"""A RNN Policy Implemented in PyTorch"""
+from rasa_core import utils
+from rasa_core.policies.policy import Policy
+from rasa_core.featurizers import TrackerFeaturizer
+from sklearn.metrics import accuracy_score
+
+import io
+import json
+import logging
+import os
+import warnings
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+logger = logging.getLogger(__name__)
+
+
+class TorchPolicy(Policy):
+    SUPPORTS_ONLINE_TRAINING = True
+
+    def __init__(self,
+                 featurizer=None,  # type: Optional[TrackerFeaturizer]
+                 model=None,  # type: Optional[tf.keras.models.Sequential]
+                 current_epoch=0,  # type: int
+                 rnn_size=32  # type: int
+                 ):
+
+        super(TorchPolicy, self).__init__(featurizer)
+
+        self.model = model
+        self.hidden_size = rnn_size
+        self.layers = 2
+        self.current_epoch = current_epoch
+
+    def model_architecture(self,
+                           input_size,  # type: Tuple[int, int]
+                           output_size  # type Tuple[int, int]
+                           ):
+
+        # assumes output_shape is size (num examples, num features)
+        rnn = RNN(input_size,
+                  self.hidden_size,
+                  output_size,
+                  self.layers,
+                  dropout=0.2)
+        return rnn
+
+    def train(self,
+              training_trackers,  # type: List[DialogueStateTracker]
+              domain,  # type: Domain
+              **kwargs  # type: Any
+              ):
+
+        training_data = self.featurize_for_training(training_trackers, domain)
+        shuffled_X, shuffled_y = training_data.shuffled_X_y()
+        batch_size, seq_len, input_size = shuffled_X.shape
+        batch_size, output_size = shuffled_y.shape
+
+        if kwargs.get('rnn_size') is not None:
+            logger.debug("Parameter rnn_size updated with "
+                         "{}".format(kwargs.get("rnn_size")))
+            self.hidden_size = kwargs.get('rnn_size')
+
+        if self.model is None:
+            self.model = self.model_architecture(input_size,
+                                                 output_size
+                                                 )
+            self.input_size = input_size
+            self.output_size = output_size
+
+        val_split = kwargs.get("validation_split", 0.1)
+        train_size = int(shuffled_X.shape[0] * val_split)
+
+        # convert to float since model requires float
+        train_X = torch.tensor(shuffled_X[:train_size]).type(torch.float)
```

Comment:
`if shuffled_X.shape[0] is 100, val_split = 0.1, train_size will be 10. Therefore train_X will have indexes [0... 10]. I am assuming this should be [0.... 90]`

Reply:
correct, just fixed it, thanks for the catch